base/compiler/ssair/passes.jl

```diff
@@ -116,36 +116,63 @@ function compute_value_for_use(ir, domtree, allblocks, du, phinodes, fidx, use_i
     end
 end
 
-function walk_to_def(compact, def, intermediaries = IdSet{Int}(), allow_phinode=true, phi_locs=Tuple{Int, Int}[])
-    if !isa(def, SSAValue)
-        return (def, 0)
+function simple_walk(compact, defssa::Union{SSAValue, NewSSAValue, OldSSAValue}, pi_callback=(pi,idx)->nothing)
+    while true
+        if isa(defssa, OldSSAValue) && already_inserted(compact, defssa)
+            rename = compact.ssa_rename[defssa.id]
+            @assert rename != defssa
+            if isa(rename, Union{SSAValue, OldSSAValue, NewSSAValue})
+                defssa = rename
+                continue
+            end
+            return rename
+        end
+        def = compact[defssa]
+        if isa(def, PiNode)
+            pi_callback(def, defssa)
+            if isa(def.val, SSAValue)
+                defssa = def.val
+            else
+                return def.val
+            end
+        elseif isa(def, Union{SSAValue, OldSSAValue, NewSSAValue})
+            defssa = def
+        elseif isa(def, Union{PhiNode, Expr})
+            return defssa
+        else
+            return def
+        end
     end
-    orig_defidx = defidx = def.id
+end
+
+function simple_walk_constraint(compact, defidx, typeconstraint = types(compact)[defidx])
+    def = simple_walk(compact, defidx, (pi,_)->(typeconstraint = typeintersect(typeconstraint, pi.typ)))
+    def, typeconstraint
+end
+
+"""
+    walk_to_defs(compact, val, intermediaries)
+
+Starting at `val` walk use-def chains to get all the leaves feeding into
+this val (pruning those leaves rules out by path conditions).
+"""
+function walk_to_defs(compact, defssa, typeconstraint, visited_phinodes=Any[])
     # Step 2: Figure out what the struct is defined as
-    def = compact[defidx]
-    typeconstraint = types(compact)[defidx]
+    def = compact[defssa]
     ## Track definitions through PiNode/PhiNode
     found_def = false
     ## Track which PhiNodes, SSAValue intermediaries
     ## we forwarded through.
-    while true
-        if isa(def, PiNode)
-            push!(intermediaries, defidx)
-            typeconstraint = typeintersect(typeconstraint, def.typ)
-            if isa(def.val, SSAValue)
-                defidx = def.val.id
-                def = compact[defidx]
-            else
-                def = def.val
-            end
-            continue
-        elseif isa(def, FastForward)
-            append!(phi_locs, def.phi_locs)
-            def = def.to
-        elseif isa(def, PhiNode)
-            # For now, we don't track setfields structs through phi nodes
-            allow_phinode || break
-            push!(intermediaries, defidx)
+    visited = IdSet{Any}()
```

Comment:
Can we use a `BitSet` here instead? One of the major features of the new representation is supposed to be that we can use fast integer sets that are just tracking indices, rather than working with values directly (as must be done in LLVM).